drivers/dai/intel/ssp/ssp.c

```diff
@@ -1927,52 +1929,35 @@ static int dai_ssp_remove(struct dai_intel_ssp *dp)
 	return 0;
 }
 
-static int dai_ssp_probe_wrapper(const struct device *dev)
+static int ssp_pm_action(const struct device *dev, enum pm_device_action action)
 {
 	struct dai_intel_ssp *dp = (struct dai_intel_ssp *)dev->data;
-	k_spinlock_key_t key;
-	int ret = 0;
-
-	key = k_spin_lock(&dp->lock);
-
-	if (dp->sref == 0) {
-		ret = dai_ssp_probe(dp);
-	}
-
-	if (!ret) {
-		dp->sref++;
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		dai_ssp_remove(dp);
+		break;
+	case PM_DEVICE_ACTION_RESUME:
+		dai_ssp_probe(dp);
+		break;
+	default:
+		return -ENOTSUP;
 	}
 
-	k_spin_unlock(&dp->lock, key);
-
-	return ret;
+	return 0;
 }
 
-static int dai_ssp_remove_wrapper(const struct device *dev)
+static int ssp_init(const struct device *dev)
 {
-	struct dai_intel_ssp *dp = (struct dai_intel_ssp *)dev->data;
-	k_spinlock_key_t key;
-	int ret = 0;
-
-	key = k_spin_lock(&dp->lock);
-
-	if (--dp->sref == 0) {
-		ret = dai_ssp_remove(dp);
-	}
-
-	k_spin_unlock(&dp->lock, key);
+	int ret;
 
+	pm_device_init_suspended(dev);
+	ret = pm_device_runtime_enable(dev);
 	return ret;
 }
 
-static int ssp_init(const struct device *dev)
-{
-	return 0;
-}
-
 static struct dai_driver_api dai_intel_ssp_api_funcs = {
-	.probe			= dai_ssp_probe_wrapper,
-	.remove			= dai_ssp_remove_wrapper,
+	.probe			= pm_device_runtime_get,
+	.remove			= pm_device_runtime_put,
```

Comment:
this API is ok for now IMHO, but we should consider removing it in future and use Zephyr PM API directly without this indirection.

Reply:
That could be, but first we'd need to move all of the drivers from sof to zephyr to do it cleanly.